BUILDIT.md

```diff
@@ -5370,7 +5376,1145 @@ when creating `timers` or `items`.
+
+The next sections use the slot feature of LiveComponents:
+
+```html
+<:column :let={metric} label="Id" key="person_id">
+  ...
+</:column>
+```
+- `<:column ...>` is a slot. It tells the `TableComponent` how to render each cell of a column. The `:let={metric}` attribute means that inside this slot, you can access each row's data with the variable metric.
+- `label="Id"` specifies the header label for this column.
+- `key="person_id"` this key will be used to sort the columns when we get there.
+
+Inside each slot, the template for rendering the cell is provided. For instance:
+
+```html
+<td class="px-6 py-4" data-test-id="person_id">
+  <a href={person_link(metric.person_id)}>
+    <%= metric.person_id %>
+  </a>
+</td>
+```
+
+This renders a table cell with the `person_id` content. The `data-test-id` attribute will be used for testing purposes, making it easier to find this specific element in test scenarios, more about that later.
+
+As a last update for this section we need to create the `is_highlighted_person?/2` method inside our `stats_live.ex` file:
+```elixir
+defmodule AppWeb.StatsLive do
+
+  ...
+
+  def is_highlighted_person?(metric, person_id),
+      do: metric.person_id == person_id
+end
+```
+
+In summary, this code replaces the static table that we had with a dynamic one, utilizing our `TableComponent`.
+
+Now you can run the code and see the same result that we had before.
+
+## 15.5 Implement Column Sorting
+
+Finally, we are in our last section to add the column sorting feature!
+
+For these we will need:
+- Add the sorting logic inside our query and method of `item.ex` file.
+- Include the sort mechanism on the Table `LiveComponent` by adding a onclick event in the table header's
+- Add the event to sort on the `Stats LiveView`
+- Create tests for everything
+
+To add the sorting logic inside `item.ex` file, we are going to create two parameters to our current method that fetches the stats. And we are going to validate the parameters as well.
+
+Open the `item.ex` file and update to the following:
+```elixir
+  ...
+
+  def person_with_item_and_timer_count(
+          sort_column \\ :person_id,
+          sort_order \\ :asc
+      ) do
+    
+    sort_column = to_string(sort_column)
+    sort_order = to_string(sort_order)
+
+    sort_column =
+      if validate_sort_column(sort_column), do: sort_column, else: "person_id"
+
+    sort_order = if validate_order(sort_order), do: sort_order, else: "asc"
+
+    sql = """
+    SELECT i.person_id,
+    COUNT(distinct i.id) AS "num_items",
+    COUNT(distinct t.id) AS "num_timers",
+    MIN(i.inserted_at) AS "first_inserted_at",
+    MAX(i.inserted_at) AS "last_inserted_at",
+    SUM(EXTRACT(EPOCH FROM (t.stop - t.start))) AS "total_timers_in_seconds"
+    FROM items i
+    LEFT JOIN timers t ON t.item_id = i.id
+    GROUP BY i.person_id
+    ORDER BY #{sort_column} #{sort_order}
```

Comment:
these should be parameterized as `$1` and `$2` to avoid `SQL injection`.

Reply:
Do you have any suggestions?

I was trying to make this work:
```
sql =
      from(i in Item,
        left_join: t in Timer,
        on: t.item_id == i.id,
        select: %{
          person_id: i.person_id,
          num_items: count(i.id, :distinct),
          num_timers: count(t.id, :distinct),
          first_inserted_at: min(i.inserted_at),
          last_inserted_at: max(i.inserted_at),
          total_timers_in_seconds: sum(fragment("EXTRACT(EPOCH FROM (? - ?))", t.stop, t.start))
        },
        group_by: i.person_id
      )

    ordered_query =
      case sort_order do
        "asc" ->
          from(q in sql, order_by: [asc: fragment("?", ^sort_column)])

        "desc" ->
          from(q in sql, order_by: [desc: fragment("?", ^sort_column)])

        _ ->
          sql
      end

    Repo.all(ordered_query)
```

but it's not working as expected, the sorting order is not working.

Reply:
@panoramix360 your `validate_sort_column/1` and `validate_order/1` functions adequately mitigate `SQL` injection. Thanks. 🙏